planner/core/rule_partition_processor.go

```diff
@@ -114,31 +124,50 @@ func (s *partitionProcessor) prune(ds *DataSource) (LogicalPlan, error) {
 		// No result after table pruning.
 		tableDual := LogicalTableDual{RowCount: 0}.Init(ds.context())
 		tableDual.schema = ds.Schema()
-		return tableDual, nil
+		return s.selectOnSomething(sel, tableDual)
 	}
 	if len(children) == 1 {
 		// No need for the union all.
-		return children[0], nil
+		return s.selectOnSomething(sel, children[0])
 	}
 	unionAll := LogicalUnionAll{}.Init(ds.context())
 	unionAll.SetChildren(children...)
 	unionAll.SetSchema(ds.schema)
-	return unionAll, nil
+	return s.selectOnSomething(sel, unionAll)
 }
 
+var solver = expression.NewPartitionPruneSolver()
+
 // canBePrune checks if partition expression will never meets the selection condition.
 // For example, partition by column a > 3, and select condition is a < 3, then canBePrune returns true.
-func (s *partitionProcessor) canBePrune(ctx sessionctx.Context, col *expression.Column, partitionCond expression.Expression, copConds []expression.Expression) (bool, error) {
-	conds := make([]expression.Expression, 0, 1+len(copConds))
+func (s *partitionProcessor) canBePrune(ctx sessionctx.Context, col *expression.Column, partitionCond expression.Expression, rootConds, copConds []expression.Expression) (bool, error) {
```

Comment:
1. how about these names:
	- s/canBePrune/canBePruned/
	- s/ctx/sctx/, to be consistent with other places
	- s/col/partCol/
	- s/partitionCond/partExpr/

2. How about record all the filters on that `DataSource` in `DataSource.PredicatePushDown()`? If so, we don't need to recollect all the filters from it and its parent, which based on the assumption that the parent of `DataSource` contains all the remained filters on that table.

3. Can we extract the partition column from the partition definition expression? If so, we can save one more parameter of this function, i.e. `col`.

Reply:
> 3. Can we extract the partition column from the partition definition expression?

`partition by to_days(datetime(x) + 1)`, the partition expression maybe complex, so it's hard to extract the partition column

Column is not useful, it's used to calculate range, but later I'll remove the calculate range logic.

Reply:
could you explain more about how to "remove the calculate range logic"?

Reply:
I'll do it in the next PR, handle the special NULL case.

Remove lines:

```
if col != nil {
    accessConds := ranger.ExtractAccessConditionsForColumn(conds, col.UniqueID)
    ....
}
```